client/client-lib/src/lib.rs

```diff
@@ -650,49 +660,74 @@ impl Client<GatewayClientConfig> {
+    /// - If the account invoice has a routing hint that refers to the current gateway,
+    ///  we return [`PaymentParameters::Internal`] to support routing as an internal payment
+    /// - If the account invoice lacks a routing hint or the hint does not match current gateway,
+    ///  we return [`PaymentParameters::Outgoing`] to support routing as an outgoing payment
     pub async fn validate_outgoing_account(
         &self,
         account: &OutgoingContractAccount,
     ) -> Result<PaymentParameters> {
         let our_pub_key = secp256k1_zkp::XOnlyPublicKey::from_keypair(&self.config.redeem_key);
 
-        if account.contract.gateway_key != our_pub_key {
+        let OutgoingContractAccount {
+            contract:
+                OutgoingContract {
+                    gateway_key,
+                    invoice,
+                    hash,
+                    timelock,
+                    ..
+                },
+            amount,
+        } = account.clone();
+
+        if gateway_key != our_pub_key {
             return Err(ClientError::NotOurKey);
         }
 
-        let invoice: Invoice = account
-            .contract
-            .invoice
-            .parse()
-            .map_err(ClientError::InvalidInvoice)?;
+        let invoice: Invoice = invoice.parse().map_err(ClientError::InvalidInvoice)?;
         let invoice_amount = Amount::from_msat(
             invoice
                 .amount_milli_satoshis()
                 .ok_or(ClientError::InvoiceMissingAmount)?,
         );
 
-        if account.amount < invoice_amount {
-            return Err(ClientError::Underfunded(invoice_amount, account.amount));
+        if amount < invoice_amount {
+            return Err(ClientError::Underfunded(invoice_amount, amount));
         }
 
-        let max_absolute_fee = account.amount - invoice_amount;
-        let max_fee_percent =
-            (max_absolute_fee.milli_sat as f64) / (invoice_amount.milli_sat as f64);
-
-        let consensus_block_height = self.context.api.fetch_consensus_block_height().await?;
-        // Calculate max delay taking into account current consensus block height and our safety
-        // margin.
-        let max_delay = (account.contract.timelock as u64)
-            .checked_sub(consensus_block_height)
-            .and_then(|delta| delta.checked_sub(self.config.timelock_delta))
-            .ok_or(ClientError::TimeoutTooClose)?;
-
-        Ok(PaymentParameters {
-            max_delay,
-            max_fee_percent,
-        })
+        // check route hint for a pubkey that could be ours
+        let maybe_our_route_hint = invoice
+            .route_hints()
+            .first()
+            .and_then(|rh| rh.0.first())
+            .map(|hop| hop.src_node_id);
+
+        if Some(self.config.node_pub_key) == maybe_our_route_hint {
+            // The route hint is indeed for this gateway. Route invoice as an internal payment
+            Ok(PaymentParameters::Internal { hash })
```

Comment:
Not a problem for a v1, but worth a TODO comment: just because there's routing hint via the GW that doesn't mean it's an internal payment, might as well be a peer of the GW LN node with a private channel. The final check should always be trying to buy the preimage from the federation. If it fails we let our LN node route.

Reply:
Makes sense. I'm planning to play around with the additional check after adding integration tests for this v1

Reply:
This turned out to be a simple change that I can include here. I need to take longer refactoring the integration tests to more realistically simulate internal vs outgoing invoices